File: PickerControls/DirectoryListing.cs
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;

namespace PickerControls
{
    public sealed class DirectoryListing : Control
    {
        public DirectoryListing()
        {
            DefaultStyleKey = typeof(DirectoryListing);
            DataContext = this;
        }

        public static DependencyProperty DirectoryProperty { get; } =
            DependencyProperty.Register("Directory", typeof(DirectoryEntry), typeof(DirectoryListing),
            new PropertyMetadata(null, OnDirectoryChangedHandler));

        public DirectoryEntry Directory
        {
            get { return GetValue(DirectoryProperty) as DirectoryEntry; }
            set { SetValue(DirectoryProperty, value); }
        }

        static void OnDirectoryChangedHandler(DependencyObject source, DependencyPropertyChangedEventArgs args)
        {
            if (args.OldValue == args.NewValue)
            {
                return;
            }
            (source as DirectoryListing).OnDirectoryChanged(args.NewValue as DirectoryEntry);
        }

        public static DependencyProperty SelectedFileProperty { get; } =
            DependencyProperty.Register("SelectedFile", typeof(FileEntry), typeof(DirectoryListing),
            new PropertyMetadata(null, OnSelectedFileChangedHandler));

        public FileEntry SelectedFile
        {
            get { return GetValue(SelectedFileProperty) as FileEntry; }
            set { SetValue(SelectedFileProperty, value); }
        }

        static void OnSelectedFileChangedHandler(DependencyObject source, DependencyPropertyChangedEventArgs args)
        {
            if (args.OldValue == args.NewValue)
            {
                return;
            }
            (source as DirectoryListing).OnSelectedFileChanged(args.NewValue as FileEntry);
        }

        DirectoryEntry directory = null;
        bool alreadyChangingDirectory = false;

        FileEntry file = null;
        bool alreadyChangingFile = false;

        void OnSelectedFileChanged(FileEntry fileEntry)
        {
            if (alreadyChangingFile)
            {
                return;
            }

            file = fileEntry;
            if (gridView != null)
            {
                gridView.SelectedItem = fileEntry;
            }
        }

        void OnDirectoryChanged(DirectoryEntry directoryEntry)
        {
            if (alreadyChangingDirectory)
            {
                return;
            }

            directory = directoryEntry;
            if (gridView != null)
            {
                AddDirectoryContents();
            }
        }

        GridView gridView;
        protected override void OnApplyTemplate()
        {
            if (gridView != null)
            {
                gridView.SelectionChanged -= GridView_SelectionChanged;
            }

            gridView = GetTemplateChild("directoryListing") as GridView;
            gridView.SelectionChanged += GridView_SelectionChanged;

            AddDirectoryContents();
        }

        void AddDirectoryContents()
        {
            if (gridView == null || directory == null)
            {
                return;
            }

            gridView.ItemsSource = directory.Files;
        }

        private void GridView_SelectionChanged(object sender, SelectionChangedEventArgs e)
        {
            if (e.AddedItems.Count <= 0)
            {
                return;
            }

            alreadyChangingFile = true;
            SelectedFile = e.AddedItems[0] as FileEntry;
            alreadyChangingFile = false;
        }
    }
}

File: PickerDesktopHosting/Control/DirectoryListing.cpp
#include "pch.h"
#include "DirectoryListing.h"
#if __has_include("DirectoryListing.g.cpp")
#include "DirectoryListing.g.cpp"
#endif

using namespace winrt::Windows::UI::Xaml;

namespace winrt::Control::implementation
{
    DependencyProperty DirectoryListing::s_SelectedFile = DependencyProperty::Register(
        winrt::hstring(L"SelectedFile"),
        winrt::xaml_typename<FileEntry>(),
        winrt::xaml_typename<winrt::Control::DirectoryListing>(),
        PropertyMetadata(nullptr, OnSelectedFileChangedHandler));

    DependencyProperty DirectoryListing::s_SelectedDirectory = DependencyProperty::Register(
        winrt::hstring(L"Directory"),
        winrt::xaml_typename<DirectoryEntry>(),
        winrt::xaml_typename<winrt::Control::DirectoryListing>(),
        PropertyMetadata(nullptr, OnDirectoryChangedHandler));

    DirectoryListing::DirectoryListing()
    {
    }

    FileEntry DirectoryListing::SelectedFile()
    {
        return m_SelectedFile;
    }

    DirectoryEntry DirectoryListing::Directory()
    {
        return m_Directory;
    }

    void DirectoryListing::Directory(DirectoryEntry directory)
    {
        m_Directory = directory;
    }

    void DirectoryListing::SelectedFile(FileEntry selectedFile)
    {
        m_SelectedFile = selectedFile;
    }

    winrt::Windows::UI::Xaml::DependencyProperty DirectoryListing::SelectedFileProperty()
    {
        return s_SelectedFile;
    }

    winrt::Windows::UI::Xaml::DependencyProperty DirectoryListing::DirectoryProperty()
    {
        return s_SelectedDirectory;
    }

    void DirectoryListing::OnSelectedFileChangedHandler(winrt::Windows::UI::Xaml::DependencyObject source, winrt::Windows::UI::Xaml::DependencyPropertyChangedEventArgs args)
    {
        if (args.OldValue() == args.NewValue())
        {
            return;
        }

        source.as<DirectoryListing>()->OnSelectedFileChanged(args.NewValue().as<FileEntry>());
    }

    void DirectoryListing::OnDirectoryChangedHandler(winrt::Windows::UI::Xaml::DependencyObject source, winrt::Windows::UI::Xaml::DependencyPropertyChangedEventArgs args)
    {
        if (args.OldValue() == args.NewValue())
        {
            return;
        }

        source.as<DirectoryListing>()->OnDirectoryChanged(args.NewValue().as<DirectoryEntry>());
    }

    void DirectoryListing::OnSelectedFileChanged(FileEntry file)
    {
        m_SelectedFile = file;
    }

    void DirectoryListing::OnDirectoryChanged(DirectoryEntry directory)
    {
        m_Directory = directory;
    }
}
